index.js

```diff
@@ -161,178 +88,249 @@ function invalidationsAsPatches(invalidated, currentTree, nextTree) {
+
+        this._logger.debug('[operation:%s] %s', operation, relativePath);
+
+        switch (operation) {
+          case 'mkdir': {
+            instrumentation.mkdir++;
+            return fs.mkdirSync(outputPath);
+          } case 'rmdir': {
+            instrumentation.rmdir++;
+            return fs.rmdirSync(outputPath);
+          } case 'unlink': {
+            instrumentation.unlink++;
+            return fs.unlinkSync(outputPath);
+          } case 'change': {
+            // wrap this in a function so it doesn't actually run yet, and can be throttled
+            let changeOperation = () => {
+              instrumentation.change++;
+              return this._handleFile(relativePath, srcDir, destDir, entry, outputFilePath, forceInvalidation, true, instrumentation);
+            };
+            if (this.async) {
+              pendingWork.push(changeOperation);
+              return;
+            }
+            return changeOperation();
+          } case 'create': {
+            // wrap this in a function so it doesn't actually run yet, and can be throttled
+            let createOperation = () => {
+              instrumentation.create++;
+              return this._handleFile(relativePath, srcDir, destDir, entry, outputFilePath, forceInvalidation, false, instrumentation);
+            };
+            if (this.async) {
+              pendingWork.push(createOperation);
+              return;
+            }
+            return createOperation();
+          } default: {
+            instrumentation.other++;
           }
-          return createOperation();
-        } default: {
-          instrumentation.other++;
         }
-      }
-    }).then(() => {
-      return queue(worker, pendingWork, this.concurrency);
-    }).then((result) => {
-      this._logger.info('applyPatches', 'duration:', timeSince(prevTime), JSON.stringify(instrumentation));
-      if (this.dependencies) {
-        this.processor.sealDependencies(this.dependencies);
-      }
-      this._needsReset = false;
-      return result;
+      }).then(() => {
+        return queue(worker, pendingWork, this.concurrency);
+      }).then((result) => {
+        this._logger.info('applyPatches', 'duration:', timeSince(prevTime), JSON.stringify(instrumentation));
+        if (this.dependencies) {
+          this.processor.sealDependencies(this.dependencies);
+        }
+        this._needsReset = false;
+        return result;
+      });
     });
-  });
-};
+  }
 
-Filter.prototype._handleFile = function(relativePath, srcDir, destDir, entry, outputPath, forceInvalidation, isChange, stats) {
-  stats.handleFile++;
+  _handleFile (relativePath, srcDir, destDir, entry, outputPath, forceInvalidation, isChange, stats) {
+    stats.handleFile++;
 
-  let handleFileStart = process.hrtime();
+    let handleFileStart = process.hrtime();
 
-  return new Promise(resolve => {
-    let result;
-    let srcPath = srcDir + '/' + relativePath;
+    return new Promise(resolve => {
```

Comment:
If we are dropping node 8 anyways, native promises get `finally`, thereby allowing us to simply use `async` functions instead. This will provide better stack traces and debugging experiences.